text/1011-object-level-security.rst

```diff
@@ -0,0 +1,320 @@
+A ``WRITE`` policy specifies a *validity* check for new or updated objects and
+affects ``INSERT`` and ``UPDATE`` expressions: if the proposed object violates
+``<expr>``, an error is raised immediately and the query is aborted.
+
+The optional ``<condition>`` expression is evaluated for every object
+affected by the statement and the policy is applied only if the expression
+evaluates to *true*.  It is essentially equivalent to joining ``<condition>``
+with ``<expr>`` with an ``AND`` operator.  The reason for a standalone clause
+is that it makes it easier to separate *when* a policy is applied from
+*how* a policy is applied.  Also, other access policies do not apply to
+``when`` expressions, but they *do* apply to ``using`` expressions.
+
+If at least one access rule is defined for a type, then all elements become
+invisible/immutable by default and must be made visible/mutable by at least
+one ``PERMIT`` rule.
+
+A ``PERMIT`` rule *adds* all objects, for which the check expression
+evaluates to true, to the set of visible objects.  ``PERMIT`` rules are
+combined using the ``OR`` operator.
+
+A ``RESTRICT`` rule restricts the set of visible objects to only objects for
+which the check expression evaluates to true.  ``RESTRICT`` rules are combined
+using the ``AND`` operator and are applied *after* all ``PERMIT`` policies.
+
+Example read policy::
+
+    type Movie {
+      property rating -> str;
+      access policy age_appropriate
+        when ((global current_user).age < 17)
+        permit read using (.rating != 'R');
+    }
+
+Example read/write policy::
+
+    type Post {
+      property author -> User;
+
+      # Only allow reading to the author, but also
+      # ensure that a user cannot set the `author` link
+      # to anything but themselves.
+      access policy author_only
+        permit all using (.author = global current_user);
+    }
+
+A combination of permit/restrict policies::
+
+    abstract type Owned {
+      link owner -> User;
+
+      # permit read access to owner
+      access policy owner_only
+        permit all using (.owner = global current_user);
+    }
+
+    abstract type Shared extending Owned {
+      # permit read access to friends
+      access policy friends_can_read
+        permit read using (global current_user in .owner.friends);
+    }
+
+    # Post inherits policies from Shared
+    # which allow access to either owner
+    # or friends initially...
+    type Post extending Shared {
+      property private -> bool;
+
+      # ... but restrict access to private posts to owner only
+      # regardless of what permissions were granted in parent types
+      access policy private_owner_only
+        when (.private)
+        restrict all using (.owner = global current_user);
+    }
+
+
+
+ALTER ACCESS POLICY
+-------------------
+
+Alter the definition of an access control policy.
+
+Required capabilities: DDL.
+
+Synopsis::
+
+    ALTER OBJECT TYPE <type-name> "{"
+        ALTER ACCESS POLICY <name>
```

Comment:
I'm not sure I understand what is intended here, with all of the arguments to `CREATE` being recapitulated in `ALTER`? Typically those are omitted and we modify the fields with subcommands, right?

Is the intent that since access policies don't have data and can't be referenced in any way, that they don't even really need `ALTER`, and we can just generate `drop`/`create` for actual changes? (Though I'd still expect ALTER to only have the name, not all the extra stuff?)

Reply:
This is meant to indicate that any part of the access policy can be altered by re-stating it in `alter`.

Reply:
Aha. Is the idea that the restatement is mandatory? And so, for example, omitting `WHEN` would drop it?

Reply:
I think it would be more consistent if we didn't actually *require* restatement.  Perhaps `drop when`?

Reply:
Yeah that could work